### DISCUSSION TITLE
Gitea에서 테스트 특기할 점

### DISCUSSION THREAD

**Discussion by Lieslot** [SWM-Codection/gitea#22] [General]
테스트 관련 주제 정리

**Comment by Lieslot** [comment]
## Gitea의 데이터베이스 테스트 시 더미 데이터를 넣는 법

Gitea의 데이터 베이스 테스트를 보면 다음과 같은 부분이 항상 맨 처음에 들어간다.

```go

assert.NoError(unittest.PrepareTestDatabase())
```

여기서 unittest는 Gitea 자체에서 구현된 부분인데 PrePareTestDatabase는 테스트 전용 데이터 베이스를 초기화하는 메소드인듯 싶다..

그리고 다음으로는 이러한 구절이 들어가는 경우가 많다.
``` go
	comment := unittest.AssertExistsAndLoadBean(t, &ai_comment.AiPullComment{ID: 2})
```
여기서 AsserExistsAndLoadBean은 데이터 베이스 테이블과 대응되는 구조체에 인자에 들어간 것과 같은 조건의 row가 존재하면 데이터 베이스 구조체를 반환하고 아니면 nil을 반환한다. 이를테면 위의 메소드는 ai_pull_comment 테이블에서 id가 2인 row가 존재하면 구조체를 반환하고 아니면 nil을 반환한다.


그런데 코드를 실행시키면 gitea에 있던 엔티티 구조체는 정상적으로 구조체를 반환하는데, 내가새롭게 추가한 테이블에 대해선 nil 값을 반환하는 문제가 있었다.

그래서 gitea에 있는 테스트 중 하나를 골라서 id = 1인 구조체에 대해 id = 100을 넣어보니 nil 값을 반환하는 것을 보고 무언가 테스트 전용 row를 넣어주는 로직이 따로 있다는 것을 직감했다. unittest.PrepareTestDatabase() 이 메소드를 타고 가다보니 **models/fixture** 디렉토리의 yml 파일과 데이터 베이스 테이블 이름을 매칭시켜서 yml 리스트 요소마다 각각 하나의 row를 데이터를 넣어주는 로직들이 있다는 것을 확인할 수 있었다. 코드가 좀 복잡한 편이라 자세한 내용은 생략하겠다.

ex)
AiPullComment라는 구조체가 있다고 할 때
파일 이름: ai_pull_comment.yml(파스칼 케이스의 대문자마다 _로 띄어서 매핑된다.)
내용:
```yml
-
  id: 2 # create 전용
  poster_id: 1
  pull_id: 1
 ```
not null이 아닌 값들은 생략하면 go의 일반적인 기본 값 규칙을 따라서 들어간다.

**Reply by cloudchamb3r** [reply]
`models/fixture/*.yml` 이 중요하네요 감사합니당 ;)
